components/brave_wallet/browser/internal/orchard_storage/orchard_types.h

```diff
@@ -0,0 +1,118 @@
+
+#include <utility>
+#include <vector>
+
+#include "base/memory/raw_ref.h"
+#include "brave/components/brave_wallet/common/zcash_utils.h"
+
+namespace brave_wallet {
+
+class ZCashOrchardStorage;
+
+// Leaf position of checkpoint.
+using CheckpointTreeState = std::optional<uint32_t>;
+
+// Checkpointed leafs are not pruned so they could be used
+// as anchors for building shielded transactions.
+// Last Orchard commitment in a block is used as a checkpoint.
+struct OrchardCheckpoint {
+  OrchardCheckpoint();
+  OrchardCheckpoint(CheckpointTreeState, std::vector<uint32_t>);
+  ~OrchardCheckpoint();
+  OrchardCheckpoint(const OrchardCheckpoint& other);
+  OrchardCheckpoint& operator=(const OrchardCheckpoint& other);
+  OrchardCheckpoint(OrchardCheckpoint&& other);
+  OrchardCheckpoint& operator=(OrchardCheckpoint&& other);
+
+  bool operator==(const OrchardCheckpoint& other) const = default;
+
+  // Leaf position of the checkpoint.
+  CheckpointTreeState tree_state_position;
+  // List of note positions that were spent at this checkpoint.
+  std::vector<uint32_t> marks_removed;
+};
+
+struct OrchardCheckpointBundle {
+  OrchardCheckpointBundle(uint32_t checkpoint_id, OrchardCheckpoint);
+  ~OrchardCheckpointBundle();
+  OrchardCheckpointBundle(const OrchardCheckpointBundle& other);
+  OrchardCheckpointBundle& operator=(const OrchardCheckpointBundle& other);
+  OrchardCheckpointBundle(OrchardCheckpointBundle&& other);
+  OrchardCheckpointBundle& operator=(OrchardCheckpointBundle&& other);
+
+  bool operator==(const OrchardCheckpointBundle& other) const = default;
+
+  // The block height serves as the checkpoint identifier.
+  uint32_t checkpoint_id = 0;
+  OrchardCheckpoint checkpoint;
+};
+
+// Address of a subtree in the shard tree.
+struct OrchardShardAddress {
+  uint8_t level = 0;
+  uint32_t index = 0;
+
+  bool operator==(const OrchardShardAddress& other) const = default;
+};
+
+// Top part of the shard tree from the root to the shard roots level
+// Used for optimization purposes in the shard tree crate.
+using OrchardShardTreeCap = std::vector<uint8_t>;
+
+// Subtree with root selected from the shard roots level.
+struct OrchardShard {
+  OrchardShard();
+  OrchardShard(OrchardShardAddress shard_addr,
+               std::optional<OrchardShardRootHash> shard_hash,
+               std::vector<uint8_t> shard_data);
+  ~OrchardShard();
+
+  OrchardShard(const OrchardShard& other);
+  OrchardShard& operator=(const OrchardShard& other);
+  OrchardShard(OrchardShard&& other);
+  OrchardShard& operator=(OrchardShard&& other);
+
+  bool operator==(const OrchardShard& other) const = default;
+
+  // Subtree root address.
+  OrchardShardAddress address;
+  // Root hash exists only on completed shards.
+  std::optional<OrchardShardRootHash> root_hash;
+  std::vector<uint8_t> shard_data;
+  // Right-most position of the subtree leaf.
+  size_t subtree_end_height = 0;
```

Comment:
This most likely should not be size_t, but speicic uint type. Blockhain state should not depend on current platform